src/rosdiscover/project/models.py

```diff
@@ -0,0 +1,135 @@
+        recovered_model_database_path = os.path.abspath("~/.rosdiscover/recovered-models")
+        self._recovered_models = RecoveredNodeModelDatabase(recovered_model_database_path)
+        self._recovery_tool = NodeRecoveryTool(app=self.config.app)
+
+    def __enter__(self) -> "ProjectModels":
+        self.open()
+        return self
+
+    def __exit__(
+        self,
+        ex_type: t.Optional[t.Type[BaseException]],
+        ex_val: t.Optional[BaseException],
+        ex_tb: t.Optional[types.TracebackType],
+    ) -> None:
+        self.close()
+
+    def open(self) -> None:
+        self._recovery_tool.open()
+
+    def close(self) -> None:
+        self._recovery_tool.close()
+
+    def _recover(self, package: str, node: str) -> t.Optional[NodeModel]:
+        # have we already recovered this model?
+        if self._recovered_models.contains(self.config, package, node):
+            return self._recovered_models.fetch(self.config, package, node)
+
+        # is this node model irrecoverable?
+        if (package, node) not in self.config.node_sources:
+            return None
+
+        # TODO we need to know the sources for this node
+        # - eventually, we want this to come from CMakeLists
+        # - for now, we can manually specify these as part of the configuration
+        sources = self.config.node_sources[(package, node)].sources
+
+        # use the recovery tool to recover the model before saving it to the database
+        model = self._recovery_tool.recover(package, node, sources)
+        self._recovered_models.store(model)
+        return model
+
+    def _fetch_handwritten(self, package: str, node: str) -> t.Optional[NodeModel]:
+        if HandwrittenModel.exists(package, node):
+            return HandwrittenModel.fetch(package, node)
+        return None
+
+    def _fetch_placeholder(self, package: str, node: str) -> NodeModel:
+        return PlaceholderModel(package, node)
+
+    def fetch(self, package: str, node: str) -> NodeModel:
+        """Retrieves the model for a given node.
+        By default, the handwritten model for this node will be returned, if
+        available, since this is likely to be the most accurate. If there is no
+        handwritten model for the given node, then the model will be statically
+        recovered from source if model recovery is allowed. Finally, if there
+        is no handwritten model and model recovery is disabled or not possible,
+        a placeholder model will be returned unless placeholders are disabled.
+
+        Parameters
+        ----------
+        package: str
+            The name of the package to which the node belongs.
+        node: str
+            The name of the type of node.
+
+        Returns
+        -------
+        NodeModel
+            When searching for a given model, highest preference will be given to
+            handwritten models. If there is no recovered model, then this method will
```

Comment:
I think you mean priority will be given to recovered models?

Reply:
For now, I've given priority to handwritten models since they're likely to be more accurate in the cases where we have them. I was thinking about adding an option to this class to switch preferences between recovered and handwritten (and maybe `allow_handwritten`, too), but I'll probably leave that to a future PR.